controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +205,114 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		infraObj, err := external.Get(ctx, c, &machine.Spec.InfrastructureRef, kcp.Namespace)
+		// Return true here because failing to get infrastructure machine should not be considered as unmatching.
+		if err != nil {
+			return true
+		}
+
+		clonedFromName, ok1 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !ok1 || !ok2 {
+			// All kcp cloned infra machines should have this annotation.
+			// Missing the annotation may be due to older version machines or adopted machines.
+			// Should not be considered as mismatch.
+			return true
+		}
+
+		//Check if infra machine's InfrastructureTemplate is matching the kcp InfrastructureTemplate.
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
+			clonedFromGroupKind != kcp.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			return false
+		}
+		return true
+	}
+}
+
+// MatchesKubernetesVersion returns a filter to find all machines that match a given Kubernetes version.
+func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		return *machine.Spec.Version == kubernetesVersion
+	}
+}
+
+// MatchesKubeadmBootstrapConfig checks if machine's KubeadmConfigSpec is equivalent with KCP's KubeadmConfigSpec.
+func MatchesKubeadmBootstrapConfig(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane, cluster clusterv1.Cluster) Func {
+	return func(machine *clusterv1.Machine) bool {
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		kcpConfigSpecLocal := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+		if bootstrapObj.Spec.InitConfiguration == nil {
+			kcpConfigSpecLocal.InitConfiguration = nil
+		}
+		if bootstrapObj.Spec.JoinConfiguration == nil {
+			kcpConfigSpecLocal.JoinConfiguration = nil
+		}
+
+		// Machines that have KubeadmClusterConfigurationAnnotation will have to match with KCP ClusterConfiguration.
+		machineClusterConfigStr, ok := machine.GetAnnotations()[controlplanev1.KubeadmClusterConfigurationAnnotation]
+		if ok {
+			machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
+			// ClusterConfiguration annotation is not correct, only solution is to rollout.
+			if err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig); err != nil {
```

Comment:
We added it as json because in the following PRs we may choose not to not a machine if some of the ClusterConfiguration fields changes, such as kube-proxy/CoreDNS. 
Since currently all changes trigger upgrade, I left it like this. But, to make it possible for future changes, chose to use json.